util/unifyfs-shell/src/unifyfs-shell.c

```diff
@@ -0,0 +1,295 @@
+{
+    if (strncmp(".quit", input, strlen(".quit")) == 0) {
+        target = -1;
+    } else if (strncmp(".target", input, strlen(".target")) == 0) {
+        char* pos = strtrim(&input[strlen(".target")]);
+        int newtarget = -1;
+
+        if (pos) {
+            newtarget = atoi(pos);
+            if (newtarget < 0 || newtarget >= env->total_ranks) {
+                printf("Target rank should be between 0 and %d\n",
+                       env->total_ranks);
+            } else {
+                target = newtarget;
+            }
+        }
+
+        printf("current target rank = %d\n", target);
+    } else if (strncmp(".help", input, strlen(".help")) == 0) {
+        unifyfs_shell_help();
+    } else {
+        printf("Unknown command %s\n", input);
+    }
+}
+
+int unifyfs_shell(unifyfs_shell_env_t* e)
+{
+    int ret = 0;
+    char* input = NULL;
+    int count = 0;
+    MPI_Status status = { 0, };
+
+    if (!e || !e->mountpoint) {
+        return -EINVAL;
+    }
+
+    env = e;
+
+    while (1) {
+        if (0 == env->rank) {
+            do {
+                input = shell_prompt();
+
+                if (input[0] == '.') {
+                    execute_builtin(input);
+                    if (target >= 0) {
+                        input = NULL;
+                    }
+                } else if (target == 0) { /* local execution on rank 0 */
+                    ret = execute_command(input, stdout);
+                    input = NULL;
+                }
+            } while (!input);
+        }
+
+        MPI_Bcast(&target, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+        if (target < 0) { /* quit the shell */
+            goto out;
+        }
+
+        if (env->rank == 0) {
+            char* outbuf = NULL;
+            unsigned long len = strlen(input) + 1;
+
+            MPI_Send(input, len, MPI_CHAR, target, 0, MPI_COMM_WORLD);
+
+            MPI_Probe(target, 0, MPI_COMM_WORLD, &status);
+            MPI_Get_count(&status, MPI_CHAR, &count);
+
+            outbuf = realloc(outbuf, count);
+            if (!outbuf) {
+                perror("failed to allocate memory");
+                ret = -errno;
+                goto out;
+            }
+
+            outbuf[count-1] = '\0';
+            MPI_Recv(outbuf, count, MPI_CHAR, target, 0,
+                     MPI_COMM_WORLD, &status);
+
+            printf("%s", outbuf);
+        } else if (env->rank == target) {
+            FILE* out = NULL;
+            char* buf = NULL;
+            size_t size = 0;
+
+            MPI_Probe(0, 0, MPI_COMM_WORLD, &status);
+            MPI_Get_count(&status, MPI_CHAR, &count);
+            MPI_Recv(cmdbuf, count, MPI_CHAR, 0, 0, MPI_COMM_WORLD, &status);
+
+            out = open_memstream(&buf, &size);
+
+            ret = execute_command(cmdbuf, out);
+
+            MPI_Send(buf, size, MPI_CHAR, 0, 0, MPI_COMM_WORLD);
+
+            fclose(out);
+        }
+
+        MPI_Barrier(MPI_COMM_WORLD);
+    }
+
+out:
```

Comment:
another case where you have an 'out' label and a function variable with the same name.